### PR TITLE
First bit of the action work.

### PR DESCRIPTION
**Pull request by m6jones**
This part enables us to do synced actions with the server. Where the server response is a hypermedia object.
Async coming in another PR.

SyncDemo

### CODE REVIEW

File: state/sirenComponents/SirenLink.js [hunk 1]
```diff
@@ -64,6 +65,7 @@ export class SirenLink {
 		if (this._token) {
 			this._childState = await stateFactory(this.link.href, shouldAttachToken(this._token, this.link));
 			this._routes.forEach((route, component) => {
+				console.log('test');
```
Comment (zinabat) [inline state/sirenComponents/SirenLink.js:68]: extraneous log statement?

File: state/store.js [hunk 2]
```diff
@@ -95,6 +95,33 @@ class StateStore {
...
+				throw response.status;
+			}
+			await this._handleCachePriming(action.token, response);
+			const json = await response.json();
+			const entity = await SirenParse(json);
+			const state = this.get(entity.getLink('self').href, action.token);
+
+			state.onServerResponse(entity);
+		} catch (err) {
+			console.log(err);
```
Comment (zinabat) [inline state/store.js:121]:
```suggestion
			console.error(err);
```

File: components/activity/name/d2l-activity-name-learning-path.js [hunk 3]
```diff
@@ -12,6 +13,7 @@ class ActivityNameLearningPath extends HypermediaLitMixin(LitElement) {
 	static get properties() {
 		return {
 			name: { type: String, observable: observableTypes.property, route: [{observable: observableTypes.link, rel: rels.specialization}]},
+			updateName: { type: Object, observable: observableTypes.action, name: 'update-name', route: [{observable: observableTypes.link, rel: rels.specialization}]},
```
Comment (zinabat) [inline components/activity/name/d2l-activity-name-learning-path.js:16]: Could this just be called `update`?

**Review by zinabat — COMMENTED** [review]
I'm so stoked about this.

File: state/sirenComponents/SirenLink.js (hunk 1 above)
Comment (m6jones) [reply state/sirenComponents/SirenLink.js:68]: I expected the linter to get these.

File: components/activity/name/d2l-activity-name-learning-path.js (hunk 3 above)
Comment (m6jones) [reply components/activity/name/d2l-activity-name-learning-path.js:16]: Given the context of the component. Yes :)

File: framework/hypermedia-lit-mixin.js [hunk 4]
```diff
@@ -54,4 +54,8 @@ export const HypermediaLitMixin = superclass => class extends superclass {
 			console.error(error);
 		}
 	}
+
+	_hasAction(action) {
+		return this[action] && this[action].has;
```
Comment (m6jones) [inline framework/hypermedia-lit-mixin.js:59]: I don't want us to check if the action exists and then to see if it has something... so thats why I made a helper.

File: state/sirenComponents/SirenAction.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+import { refreshToken } from '../token.js';
+
+export class SirenAction {
+	static basicInfo({name: id, token}) {
+		return { id, token };
+	}
+
+	constructor({id: name, token}) {
+		this._components = new Component();
+		this._action = { has: false, perform: () => undefined };
```
Comment (m6jones) [inline state/sirenComponents/SirenAction.js:12]: These names probably need to be changed but lets do that when actions have the state and better fit all use cases. This is one of 3.

File: state/store.js [hunk 6]
```diff
@@ -95,6 +95,33 @@ class StateStore {
 		}
 	}
 
+	async performAction(action, input) {
```
Comment (m6jones) [inline state/store.js:98]: `fetch` and `performAction` are the same method. Maybe if we make `fetch` less specific they can finally be one!

Might need a interface for objects that can be fetched.

File: state/store.js [hunk 7]
```diff
@@ -95,6 +95,33 @@ class StateStore {
...
+
+		const headers = action.header();
+		!action.token.cookie && headers.set('Authorization', `Bearer ${action.token.value}`);
+
+		const fetch = !action.token.cookie
+			? this._d2lfetch
+			: this._d2lfetch.removeTemp('auth');
+
+		try {
+			const response = await fetch.fetch(href, { headers, body, method: action.method });
```
Comment (zinabat) [inline state/store.js:110]: lol, `fetch.fetch()` maybe rename to `d2lfetch`?

**Review by zinabat — APPROVED** [review]
We need this for assignment editor

**Pull request merged**